README.md
<p align="center"><img src="screenshots/screencast.gif"></p>

<h2 align="center">External Note Editor for the Browser for Anki</h2>

<p align="center">
<a title="Latest (pre-)release" href="https://github.com/glutanimate/browser-external-note-editor/releases"><img src ="https://img.shields.io/github/release-pre/glutanimate/browser-external-note-editor.svg?colorB=brightgreen"></a>
<a title="License: GNU AGPLv3" href="https://github.com/glutanimate/browser-external-note-editor/blob/master/LICENSE"><img  src="https://img.shields.io/badge/license-GNU AGPLv3-green.svg"></a>
<a title="Rate on AnkiWeb" href="https://ankiweb.net/shared/info/1996266159"><img src="https://glutanimate.com/logos/ankiweb-rate.svg"></a>
<br>
<a title="Buy me a coffee :)" href="https://ko-fi.com/X8X0L4YV"><img src="https://img.shields.io/badge/ko--fi-contribute-%23579ebd.svg"></a>
<a title="Support me on Patreon :D" href="https://www.patreon.com/bePatron?u=7522179"><img src="https://img.shields.io/badge/patreon-support-%23f96854.svg"></a>
<a title="Follow me on Twitter" href="https://twitter.com/intent/user?screen_name=glutanimate"><img src="https://img.shields.io/twitter/follow/glutanimate.svg"></a>
</p>

> Comfortably browse through and edit your Anki cards

This is an add-on for the spaced-repetition flashcard app [Anki](https://apps.ankiweb.net/) that adds the ability to edit your flashcards in a separate window from Anki's card browser.

### Table of Contents <!-- omit in toc -->

<!-- MarkdownTOC levels="1,2,3" -->

- [Installation](#installation)
- [Documentation](#documentation)
- [Building](#building)
- [Contributing](#contributing)
- [License and Credits](#license-and-credits)

<!-- /MarkdownTOC -->

### Installation

#### AnkiWeb <!-- omit in toc -->

The easiest way to install External Note Editor for the Browser is through [AnkiWeb](https://ankiweb.net/shared/info/1996266159).

#### Manual installation <!-- omit in toc -->

Please click on the entry corresponding to your Anki version:

<details>

<summary><i>Anki 2.1</i></summary>

1. Make sure you have the [latest version](https://apps.ankiweb.net/#download) of Anki 2.1 installed. Earlier releases (e.g. found in various Linux distros) do not support `.ankiaddon` packages.
2. Download the latest `.ankiaddon` package from the [releases tab](https://github.com/glutanimate/browser-external-note-editor/releases) (you might need to click on *Assets* below the description to reveal the download links)
3. From Anki's main window, head to *Tools* → *Add-ons*
4. Drag-and-drop the `.ankiaddon` package onto the add-ons list
5. Restart Anki

Video summary:

<img src="https://raw.githubusercontent.com/glutanimate/docs/master/anki/add-ons/media/ankiaddon-installation.gif" width=640>

</details>

### Documentation

For further information on the use of this add-on please check out [the description text](docs/description.md) for AnkiWeb.

### Building

With [Anki add-on builder](https://github.com/glutanimate/anki-addon-builder/) installed:

    git clone https://github.com/glutanimate/browser-external-note-editor.git
    cd browser-external-note-editor
    aab build

For more information on the build process please refer to [`aab`'s documentation](https://github.com/glutanimate/anki-addon-builder/#usage).

### Contributing

Contributions are welcome! Please review the [contribution guidelines](./CONTRIBUTING.md) on how to:

- Report issues
- File pull requests
- Support the project as a non-developer

### License and Credits

*External Note Editor for the Browser* is *Copyright © 2017-2020 [Aristotelis P.](https://glutanimate.com/) (Glutanimate)*

External Note Editor for the Browser is free and open-source software. The add-on code that runs within Anki is released under the GNU AGPLv3 license, extended by a number of additional terms. For more information please see the [LICENSE](https://github.com/glutanimate/browser-external-note-editor/blob/master/LICENSE) file that accompanied this program.

Please note that this program uses the [Libaddon](https://github.com/glutanimate/anki-libaddon/) library which comes with [its own additional terms extending the GNU AGPLv3 license](https://github.com/glutanimate/anki-libaddon/blob/master/LICENSE). You may only copy, distribute, or modify the present compilation of this program with Libaddon under the combined licensing terms specified by both licenses.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY.


----

<b>
<div align="center">The continued development of this add-on is made possible <br>thanks to my <a href="https://www.patreon.com/glutanimate">Patreon</a> and <a href="https://ko-fi.com/X8X0L4YV">Ko-Fi</a> supporters.
<br>You guys rock ❤️ !</div>
</b>

CHANGELOG.md
# Changelog

All notable changes to External Note Editor for the Browser will be documented here. You can click on each release number to be directed to a detailed log of all code commits for that particular release. The download links will direct you to the GitHub release page, allowing you to manually install a release if you want.

If you enjoy External Note Editor for the Browser, please consider supporting my work on Patreon, or by buying me a cup of coffee :coffee::

<p align="center">
<a href="https://www.patreon.com/glutanimate" rel="nofollow" title="Support me on Patreon 😄"><img src="https://glutanimate.com/logos/patreon_button.svg"></a>      <a href="https://ko-fi.com/X8X0L4YV" rel="nofollow" title="Buy me a coffee 😊"><img src="https://glutanimate.com/logos/kofi_button.svg"></a>
</p>

:heart: My heartfelt thanks goes out to everyone who has supported this add-on through their tips, contributions, or any other means (you know who you are!). All of this would not have been possible without you. Thank you for being awesome!

## [Unreleased]

## [1.0.0] - 2020-03-31

### [Download](https://github.com/glutanimate/browser-external-note-editor/releases/tag/v1.0.0)

### Added

- Anki 2.1 support! (thanks to @ijgnd for his help with this!)
- Customizable hotkey

### Fixed

- Various smaller issues with dialog management

### Changed

- Rewrote major parts of the add-on, implement brand-new changes to the Anki add-on API (which should make the add-on future-proof for the Anki updates to come)

## [1.0.0-beta.1] - 2020-03-10

### Added

- Anki 2.1 support!
- Customizable hotkey

### Fixed

- Various smaller issues with dialog management

### Changed

- Rewrote major parts of the add-on, implement brand-new changes to the Anki add-on API (which should make the add-on future-proof for the Anki updates to come)

## 0.1.0 - 2017-02-21

### Added

- Initial release of External Note Editor for the Browser

[1.0.0-beta.1]: https://github.com/glutanimate/puppy-reinforcement/releases/tag/v1.0.0-beta.1

-----

The format of this file is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

src/browser_external_note_editor/config.py
# -*- coding: utf-8 -*-

# External Note Editor for the Browser Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Handles add-on configuration
"""

from aqt import mw

from .libaddon.anki.configmanager import ConfigManager

config = ConfigManager(mw)


src/browser_external_note_editor/editor.py
# -*- coding: utf-8 -*-

# External Note Editor for the Browser Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

from PyQt5.QtWidgets import QDialog, QDialogButtonBox
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QKeySequence

import aqt.editor
from aqt.forms import editcurrent
from aqt.main import AnkiQt
from aqt.utils import saveGeom, restoreGeom
from anki.hooks import addHook, remHook

from anki.utils import isMac
from anki.lang import _

from aqt.browser import Browser
from aqt.editcurrent import EditCurrent
from aqt import dialogs

try:
    from aqt import gui_hooks

    NEW_STYLE_HOOKS = True
except ImportError:
    NEW_STYLE_HOOKS = False


class BrowserEditCurrent(QDialog):
    """Based on editcurrent.EditCurrent"""

    def __init__(self, mw: AnkiQt, browser: Browser):
        super().__init__(None, Qt.Window)
        mw.setupDialogGC(self)
        self.mw = mw
        self.browser = browser
        self.form = aqt.forms.editcurrent.Ui_Dialog()
        self.form.setupUi(self)
        self.setWindowTitle(_("Edit Current"))
        self.setMinimumHeight(400)
        self.setMinimumWidth(250)
        self.form.buttonBox.button(QDialogButtonBox.Close).setShortcut(
            QKeySequence("Ctrl+Return")
        )
        self.editor = aqt.editor.Editor(self.mw, self.form.fieldsArea, self)
        self.editor.card = self.browser.card
        if self.browser.card is not None:
            self.editor.setNote(self.browser.card.note(), focusTo=0)
        restoreGeom(self, "browsereditcurrent")
        if NEW_STYLE_HOOKS:
            gui_hooks.state_did_reset.append(self.onReset)
        else:
            addHook("reset", self.onReset)
        self.mw.requireReset()
        self.show()
        # reset focus after open, taking care not to retain webview
        # pylint: disable=unnecessary-lambda
        self.mw.progress.timer(100, lambda: self.editor.web.setFocus(), False)

    def onReset(self):
        # lazy approach for now: throw away edits
        try:
            n = self.browser.card.note()
            n.load()
        except:
            # card's been deleted
            if NEW_STYLE_HOOKS:
                gui_hooks.state_did_reset.remove(self.onReset)
            else:
                remHook("reset", self.onReset)
            # remHook("remNotes", self.onRemoveNotes)
            self.editor.setNote(None)
            self.mw.reset()
            aqt.dialogs.markClosed("BrowserEditCurrent")
            self.close()
            return
        self.editor.setNote(n)

    def onSave(self):
        remHook("reset", self.onReset)
        self.editor.saveNow()
        self.browser.externalNid = None
        self.browser.form.splitter.widget(1).setVisible(True)
        self.browser.editor.setNote(self.browser.card.note(reload=True))
        saveGeom(self, "browsereditcurrent")
        aqt.dialogs.close("BrowserEditCurrent")

    def onRemoveNotes(self, ids):
        """Close window before shown note is deleted"""
        """Might not be necessary because it's handled by onReset"""
        # called via addHook("remNotes") or new-style hook
        # if my id in ids:
        # self.close

    def canClose(self):
        return True

    def reject(self):
        self.saveAndClose()

    def saveAndClose(self):
        self.editor.saveNow(self._saveAndClose)

    def _saveAndClose(self) -> None:
        if NEW_STYLE_HOOKS:
            gui_hooks.state_did_reset.remove(self.onReset)
        else:
            remHook("reset", self.onReset)

        saveGeom(self, "browsereditcurrent")
        aqt.dialogs.markClosed("BrowserEditCurrent")
        super().reject()

    def closeWithCallback(self, onsuccess):
        def callback():
            self._saveAndClose()
            onsuccess()

        self.editor.saveNow(callback)


def initialize_editor():
    # Register new dialog in DialogManager:
    dialogs._dialogs["BrowserEditCurrent"] = [BrowserEditCurrent, None]


src/browser_external_note_editor/browser.py
# -*- coding: utf-8 -*-

# External Note Editor for the Browser Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

import sip

import aqt.editor
from anki.hooks import addHook
from aqt.browser import Browser
from aqt.qt import *
from aqt.utils import tooltip

from .config import config


def hide_browser_editor(browser: Browser):
    if sip.isdeleted(browser) or not browser.editor:
        return
    browser.form.splitter.widget(1).setVisible(False)
    browser.editor.setNote(None)


def show_browser_editor(browser: Browser):
    if sip.isdeleted(browser) or not browser.editor or not browser.card:
        return
    browser.form.splitter.widget(1).setVisible(True)
    browser.editor.setNote(browser.card.note(reload=True))


def _on_edit_window(browser: Browser) -> bool:
    """Launch BrowserEditCurrent instance"""
    cids = browser.selectedCards()
    if not cids or not browser.card:
        tooltip("No cards selected")
        return False
    elif len(cids) > 1:
        tooltip("Please select just one card")
        return False

    hide_browser_editor(browser)

    browser.external_editor = aqt.dialogs.open(
        "BrowserEditCurrent", browser.mw, browser, browser.card
    )

    return True


def _on_setup_menus(browser):
    """Create menu entry and set attributes up"""
    menu = browser.form.menuEdit
    menu.addSeparator()
    a = menu.addAction("Edit in New Window")
    a.setShortcut(QKeySequence(config["local"]["hotkey"]))
    a.triggered.connect(lambda _, o=browser: _on_edit_window(o))
    browser.external_editor = None


def initialize_browser():
    try:  # >= 2.1.20
        from aqt.gui_hooks import browser_menus_did_init

        browser_menus_did_init.append(_on_setup_menus)
    except (ImportError, ModuleNotFoundError, AttributeError):
        addHook("browser.setupMenus", _on_setup_menus)


src/browser_external_note_editor/__init__.py
# -*- coding: utf-8 -*-

# External Note Editor for the Browser Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Initializes add-on components.
"""


from ._version import __version__  # noqa: F401

from .libaddon import maybeVendorTyping

maybeVendorTyping()


def initialize_addon():
    """Initializes add-on after performing a few checks
    
    Allows more fine-grained control over add-on execution, which can
    be helpful when implementing workarounds for Anki bugs (e.g. the module
    import bug present in all Anki 2.1 versions up to 2.1.14)
    """

    from .libaddon import checkFor2114ImportError
    from .consts import ADDON

    if not checkFor2114ImportError(ADDON.NAME):
        return False

    from .libaddon.consts import setAddonProperties

    setAddonProperties(ADDON)

    # from .libaddon.debug import maybeStartDebugging

    # maybeStartDebugging()

    from .browser import initialize_browser
    from .editcurrent_browser import initialize_editcurrent_browser

    initialize_browser()
    initialize_editcurrent_browser()


initialize_addon()


src/browser_external_note_editor/_version.py
# -*- coding: utf-8 -*-

# External Note Editor for the Browser Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Version information
"""

__version__ = "1.0.0"


src/browser_external_note_editor/editcurrent_browser.py
# -*- coding: utf-8 -*-

# External Note Editor for the Browser Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

from typing import List

from PyQt5.QtCore import Qt
from PyQt5.QtGui import QKeySequence
from PyQt5.QtWidgets import QDialog, QDialogButtonBox

import aqt
from anki.cards import Card
from anki.hooks import addHook, remHook
from anki.lang import _
from anki.storage import _Collection
from aqt.browser import Browser
from aqt.main import AnkiQt
from aqt.utils import restoreGeom, saveGeom, tooltip

from .browser import hide_browser_editor, show_browser_editor


class BrowserEditCurrent(QDialog):
    """Based on editcurrent.EditCurrent"""

    def __init__(self, mw: AnkiQt, browser: Browser, card: Card):
        super().__init__(None, Qt.Window)
        mw.setupDialogGC(self)
        self.mw = mw
        self.browser = browser
        self.card = card
        self.form = aqt.forms.editcurrent.Ui_Dialog()
        self.form.setupUi(self)
        self.setWindowTitle(_("Edit Current"))
        self.setMinimumHeight(400)
        self.setMinimumWidth(250)
        self.form.buttonBox.button(QDialogButtonBox.Close).setShortcut(
            QKeySequence("Ctrl+Return")
        )
        self.editor = aqt.editor.Editor(self.mw, self.form.fieldsArea, self)
        self.editor.card = card
        self.editor.setNote(card.note(), focusTo=0)
        restoreGeom(self, "browsereditcurrent")
        self.mw.requireReset()
        self._setupHooks()
        self.show()
        # reset focus after open, taking care not to retain webview
        # pylint: disable=unnecessary-lambda
        self.mw.progress.timer(100, lambda: self.editor.web.setFocus(), False)

    # Qt API

    def reject(self):
        self.saveAndClose()

    # Setup

    ## Hooks

    def _setupHooks(self):
        # TODO: close on browser close?
        try:
            from aqt.gui_hooks import state_did_reset, browser_did_change_row
            from anki.hooks import notes_will_be_deleted

            state_did_reset.append(self._onReset)
            notes_will_be_deleted.append(self._onNotesWillBeDeleted)
            browser_did_change_row.append(self._onBrowserRowChanged)
        except (ImportError, ModuleNotFoundError, AttributeError):
            addHook("reset", self._onReset)
            addHook("remNotes", self._onNotesWillBeDeleted)
            addHook("browser.rowChanged", self._onBrowserRowChanged)  # requires >=2.1.5

    def _teardownHooks(self):
        try:
            from aqt.gui_hooks import state_did_reset, browser_did_change_row
            from anki.hooks import notes_will_be_deleted

            state_did_reset.remove(self._onReset)
            notes_will_be_deleted.remove(self._onNotesWillBeDeleted)
            browser_did_change_row.remove(self._onBrowserRowChanged)
        except (ImportError, ModuleNotFoundError, AttributeError):
            remHook("reset", self._onReset)
            remHook("remNotes", self._onNotesWillBeDeleted)
            remHook("browser.rowChanged", self._onBrowserRowChanged)  # requires >=2.1.5

    ## Hook subscribers

    def _onBrowserRowChanged(self, _: Browser):
        if not self.editor.note:
            return
        selected_nids = self.browser.selectedNotes()
        if selected_nids and selected_nids[0] == self.editor.note.id:
            hide_browser_editor(self.browser)

    def _onNotesWillBeDeleted(self, col: _Collection, nids: List[int]):
        if not self.editor.note:
            return
        current_nid = self.editor.note.id
        if current_nid in nids:
            self.saveAndClose()

    def _onReset(self):
        # lazy approach for now: throw away edits
        try:
            n = self.card.note()
            n.load()
        except Exception as e:  # noqa: E722
            print(e)
            # card's been deleted
            self._teardownHooks()
            # remHook("remNotes", self.onRemoveNotes)
            self.editor.setNote(None)
            self.mw.reset()
            aqt.dialogs.markClosed("BrowserEditCurrent")
            self.close()
            return
        self.editor.setNote(n)

    def saveAndClose(self):
        self.editor.saveNow(self._saveAndClose)

    def _saveAndClose(self) -> None:
        self._teardownHooks()
        if self.browser:
            show_browser_editor(self.browser)
            self.browser.edit_current = None
        saveGeom(self, "browsereditcurrent")
        aqt.dialogs.markClosed("BrowserEditCurrent")
        super().reject()

    # DialogManager API

    def reopen(self, mw: AnkiQt, browser: Browser, card: Card):
        self.editor.saveNow(lambda: self._reopen(mw, browser, card))

    def _reopen(self, mw: AnkiQt, browser: Browser, card: Card):
        self.mw = mw
        self.browser = browser
        self.card = card
        self.editor.card = card
        self.editor.setNote(card.note(), focusTo=0)
        self._onReset()

    def closeWithCallback(self, onsuccess):
        def callback():
            self._saveAndClose()
            onsuccess()

        self.editor.saveNow(callback)


def initialize_editcurrent_browser():
    # Register new dialog in DialogManager:
    try:  # >= 2.1.22
        aqt.dialogs.register_dialog("BrowserEditCurrent", BrowserEditCurrent, None)
    except AttributeError:
        aqt.dialogs._dialogs["BrowserEditCurrent"] = [BrowserEditCurrent, None]


src/browser_external_note_editor/consts.py
# -*- coding: utf-8 -*-

# External Note Editor for the Browser Add-on for Anki
#
# Copyright (C) 2017-2020  Aristotelis P. <https://glutanimate.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version, with the additions
# listed at the end of the license file that accompanied this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# NOTE: This program is subject to certain additional terms pursuant to
# Section 7 of the GNU Affero General Public License.  You should have
# received a copy of these additional terms immediately following the
# terms and conditions of the GNU Affero General Public License that
# accompanied this program.
#
# If not, please request a copy through one of the means of contact
# listed here: <https://glutanimate.com/contact/>.
#
# Any modifications to this file must keep this entire header intact.

"""
Addon-wide constants
"""

from ._version import __version__

try:
    from .data.patrons import MEMBERS_CREDITED, MEMBERS_TOP
except ImportError:
    MEMBERS_CREDITED = MEMBERS_TOP = ()


__all__ = [
    "ADDON"
]


class ADDON:
    """Class storing general add-on properties
    Property names need to be all-uppercase with no leading underscores
    """
    NAME = "External Note Editor for the Browser"
    MODULE = "browser_external_note_editor"
    ID = "1996266159"
    VERSION = __version__
    LICENSE = "GNU AGPLv3"
    AUTHORS = (
        {"name": "Aristotelis P. (Glutanimate)", "years": "2017-2020",
         "contact": "Glutanimate"},
    )
    AUTHOR_MAIL = "ankiglutanimate [αt] gmail . com"
    LIBRARIES = ()
    CONTRIBUTORS = ()
    SPONSORS = ()
    MEMBERS_CREDITED = MEMBERS_CREDITED
    MEMBERS_TOP = MEMBERS_TOP
    LINKS = {
        "patreon": "https://www.patreon.com/glutanimate",
        "bepatron": "https://www.patreon.com/bePatron?u=7522179",
        "coffee": "http://ko-fi.com/glutanimate",
        "description": "https://ankiweb.net/shared/info/{}".format(ID),
        "rate": "https://ankiweb.net/shared/review/{}".format(ID),
        "twitter": "https://twitter.com/glutanimate",
        "youtube": "https://www.youtube.com/c/glutanimate",
        "help": ""
    }
